beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,720 @@
+        verify_propagation_slot_range(chain, contribution)?;
+
+        // Validate subcommittee index.
+        if contribution.subcommittee_index >= SYNC_COMMITTEE_SUBNET_COUNT {
+            return Err(Error::InvalidSubcommittee {
+                subcommittee_index: contribution.subcommittee_index,
+                subcommittee_size: SYNC_COMMITTEE_SUBNET_COUNT,
+            });
+        }
+
+        // Ensure the valid sync contribution has not already been seen locally.
+        let contribution_root = contribution.tree_hash_root();
+        if chain
+            .observed_sync_contributions
+            .write()
+            .is_known(contribution, contribution_root)
+            .map_err(|e| Error::BeaconChainError(e.into()))?
+        {
+            return Err(Error::SyncContributionAlreadyKnown(contribution_root));
+        }
+
+        // Ensure there has been no other observed aggregate for the given `aggregator_index`.
+        //
+        // Note: do not observe yet, only observe once the sync contribution has been verified.
+        let observed_key =
+            SlotSubcommitteeIndex::new(contribution.slot, contribution.subcommittee_index);
+        match chain
+            .observed_sync_aggregators
+            .read()
+            .validator_has_been_observed(observed_key, aggregator_index as usize)
+        {
+            Ok(true) => Err(Error::AggregatorAlreadyKnown(aggregator_index)),
+            Ok(false) => Ok(()),
+            Err(ObservedAttestersError::ValidatorIndexTooHigh(i)) => {
+                Err(Error::ValidatorIndexTooHigh(i))
+            }
+            Err(e) => Err(BeaconChainError::from(e).into()),
+        }?;
+
+        // Ensure the block being voted for (contribution.beacon_block_root) passes validation.
+        //
+        // This indirectly checks to see if the `contribution.beacon_block_root` is in our fork
+        // choice. Any known, non-finalized, processed block should be in fork choice, so this
+        // check immediately filters out sync committee contributions that attest to a block that
+        // has not been processed.
+        //
+        // Sync committee contributions must be for a known block. If the block is unknown, we
+        // simply drop the sync committee contribution and do not delay consideration for later.
+        verify_head_block_is_known(chain, contribution.beacon_block_root)?;
+
+        // Ensure that the sync committee message has participants.
+        if contribution.aggregation_bits.is_zero() {
+            return Err(Error::EmptyAggregationBitfield);
+        }
+
+        // Ensure the aggregator's pubkey is in the declared subcommittee of the current sync committee
+        let pubkey_bytes = chain
+            .validator_pubkey_bytes(aggregator_index as usize)?
+            .ok_or(Error::UnknownValidatorIndex(aggregator_index as usize))?;
+        let sync_subcommittee_pubkeys = chain
+            .head_sync_committee_next_slot()?
```

Comment:
I just ran into an issue with this validation condition on a network with partial participation: if there's a skipped slot in the last slot of a sync committee period, e.g. slot 63, then the head will still be sitting at slot 62, and `head_sync_committee_next_slot` will consider slot 63 (period 0) rather than slot 64 (period 1). I think we should replace `head_sync_committee_next_slot` with a function that takes a `slot` argument so that it isn't possible for this mixup to occur.

Arguably this is a flaw in the P2P spec, which doesn't define what it's `state` variable is meant to be (presumably the head state advanced to the current slot):

> [REJECT] The aggregator's validator index is in the declared subcommittee of the current sync committee -- i.e. `state.validators[contribution_and_proof.aggregator_index].pubkey in get_sync_subcommittee_pubkeys(state, contribution.subcommittee_index)`.

A test for this case would also be nice, but I'm planning to write one at the HTTP API level, so happy to avoid duplicating the effort.

Reply:
Nice find... If the issue is that we are using a state that hasn't been advanced though, wouldn't we have the same issue with `state.current_sync_committee` and `state.next_sync_committee` potentially not being updated?  This is making me wonder if we should be using the `snapshot_cache` rather than the `canonical_head`. 

I'm also seeing in signature verification we are getting the `state.fork()` off the canonical head... so if there are skip slots at the `altair` fork boundary would that too be out of date?

Reply:
Oh yeah, you're right! There are probably a wealth of issues at the fork boundary in the presence of skip slots 🤔 

For `state.fork()` we could use the pure function that the VC uses, which maps a ChainSpec and a slot to a Fork. Although that doesn't solve the issue of the sync committees not being available due to the state not being transitioned yet (Base -> Altair). The snapshot cache could be useful in some cases, although the state advance service will only advance by one slot, so we still need another way of dealing with two skip slots before the fork boundary. Maybe we need something like the `state_advance_for_committee_caches` used in attestation processing that has a fast path for the common case and falls back to doing expensive reads/clones/state processing for full generality.

We can dig into it some more next week

Reply:
I think I made good progress on this today. A bunch of the errors seem to be handled by this change: https://github.com/sigp/lighthouse/commit/503c60d2c7eb3acbb669e233a92bd331b693f381

I'm still missing something about the aggregator indexes that I'll need to dig into tomorrow. Once I finish that I'll make a cleaner commit that we can cherry-pick on to this branch.

Reply:
Nice man! Those changes look good

Reply:
I've pushed them to this branch now, I think in a way that will minimise conflicts upstream :crossed_fingers: